crates/mysten-network/src/server.rs

```diff
@@ -26,22 +30,42 @@ use tower::{
     util::Either,
     Service, ServiceBuilder,
 };
+use tower_http::classify::{GrpcErrorsAsFailures, SharedClassifier};
+use tower_http::propagate_header::PropagateHeaderLayer;
+use tower_http::set_header::SetRequestHeaderLayer;
+use tower_http::trace::{DefaultMakeSpan, TraceLayer};
 
-pub struct ServerBuilder {
-    router: Router<WrapperService>,
+pub struct ServerBuilder<M: MetricsCallbackProvider = DefaultMetricsCallbackProvider> {
+    router: Router<WrapperService<M>>,
     health_reporter: tonic_health::server::HealthReporter,
 }
 
-type WrapperService = Stack<
+type AddPathToHeaderFunction = fn(&Request<Body>) -> Option<HeaderValue>;
+
+type WrapperService<M> = Stack<
     Stack<
-        Either<LoadShedLayer, Identity>,
-        Stack<Either<GlobalConcurrencyLimitLayer, Identity>, Identity>,
+        PropagateHeaderLayer,
+        Stack<
+            TraceLayer<
+                SharedClassifier<GrpcErrorsAsFailures>,
+                DefaultMakeSpan,
+                MetricsHandler<M>,
+                MetricsHandler<M>,
+            >,
+            Stack<
+                SetRequestHeaderLayer<AddPathToHeaderFunction>,
+                Stack<
+                    Either<LoadShedLayer, Identity>,
+                    Stack<Either<GlobalConcurrencyLimitLayer, Identity>, Identity>,
+                >,
+            >,
+        >,
     >,
     Identity,
 >;
```

Comment:
I probably need to get around to making this generic so that we can eliminate this super large type alias